nixops/resources/output.py

```diff
@@ -0,0 +1,110 @@
+# -*- coding: utf-8 -*-
+
+# Arbitrary JSON.
+
+import os
+import nixops.util
+import nixops.resources
+
+import tempfile
+import shutil
+import subprocess
+import hashlib
+
+
+# For typing
+from nixops.deployment import Deployment
+from xml.etree.ElementTree import Element
+from nixops.nix_expr import Function
+from typing import Optional, List, Dict, Tuple
+
+class OutputDefinition(nixops.resources.ResourceDefinition):
+    """Definition of an Output."""
+
+    @classmethod
+    def get_type(cls):
+        # type: () -> (str)
+        return "output"
+
+    @classmethod
+    def get_resource_type(cls):
+        # type: () -> (str)
+        return "output"
+
+    def __init__(self, xml):
+        # type: (Element) -> None
+        nixops.resources.ResourceDefinition.__init__(self, xml)
+        temp = xml.find("attrs/attr[@name='script']/string")
+        self.script = temp.get("value") if temp is not None else ""
+        temp = xml.find("attrs/attr[@name='name']/string")
+        self.name = temp.get("value") if temp is not None else ""
+
+    def show_type(self):
+        # type: () -> (str)
+        return "{0}".format(self.get_type())
+
+class OutputState(nixops.resources.ResourceState):
+    """State of an output."""
+    state = nixops.util.attr_property("state", nixops.resources.ResourceState.MISSING, int)
+    script = nixops.util.attr_property("script",None)
+    value = nixops.util.attr_property("value",None)
+    name =  nixops.util.attr_property("name",None)
+
+    @classmethod
+    def get_type(cls):
+        # type: () -> (str)
+        return "output"
+
+    @property
+    def resource_id(self):
+        # type: () -> Optional[str]
+        if self.value is not None:
+            # Avoid printing any potential secret information
+            return "{0}-{1}".format(self.name,hashlib.sha256(self.value).hexdigest()[:32])
+        else:
+            return None
+
+    def __init__(self, depl, name, id):
+        # type: (Deployment,str,int) -> None
+        self.id = id
+        nixops.resources.ResourceState.__init__(self, depl, name, id)
+
+    def create(self, defn, check, allow_reboot, allow_recreate):
+        # type: (OutputDefinition,bool,bool,bool) -> None
+        if (defn.script is not None ) and (self.script != defn.script) or self.value is None:
+            self.name = defn.name
+            try:
+                output_dir = tempfile.mkdtemp(prefix="nixops-output-tmp")
+                self.log("Running shell function for output ‘{0}’...".format(defn.name))
+                env = {} # type: Dict[str,str]
+                env.update(os.environ)
+                env.update({"out":output_dir})
+                res = subprocess.check_output(
+                        [defn.script],
+                        env=env,
+                        shell=True)
```

Comment:
Please note that this will make the script dependent on the system's shell. For example if you're using NixOps on a Debian machine, you'll end up getting `dash` instead of `bash`. So I'd probably wrap the script with `stdenv.shell` or `pkgs.bash` so that if the `script` is using bashisms it will still work.

One possibility would be to add another option to the resource, like eg. `executable`, which defaults to using something like eg. `pkgs.writeScript "somename" "#!${pkgs.stdenv.shell}\n${config.script}"`. That way you only need to run the `executable` here without the need to set `shell=True`.

Reply:
Incorporated all the other feedback. I'm not sure how to do this one. Whenever I try to do this, it seems the writeScript gets evaluated, but not instantiated.

Reply:
Can you post an output of that with `--debug`?